### PR TITLE
Node authorizer enhancement: using kubelet's client can only view resources belonging to its own node

### LINKED ISSUE #119470: The [client certificate used by] kubelet can view all node and pod resources
### What would you like to be added?

In the current node authentication mode, the client certificate used by kubelet can view all node and pod resources. Once the node is compromised, the cluster information will be leaked.

### Why is this needed?

using kubelet's client can only view resources belonging to its own node

### PR DESCRIPTION
**Pull request by sunnoy**
Node authorizer enhancement: using kubelet's client certificate can only view resources belonging to its own node

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

In the current node authentication mode, the client certificate used by kubelet can view all node and pod resources. Once the node is compromised, the cluster information will be leaked.

so

Node authorizer enhancement: using kubelet's client can only view resources belonging to its own node

#### Which issue(s) this PR fixes:

Fixes #119470

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Review by sftim — COMMENTED** [review]
This PR _does_ introduce a user facing change. We should only make it:
- behind a feature gate
- with a changelog entry

You may need to write an [enhancement request](https://github.com/kubernetes/enhancements/) to track the improvement here.

**Comment by liggitt** [comment]
> This PR _does_ introduce a user facing change. We should only make it:
> 
> * behind a feature gate
> * with a changelog entry
> 
> You may need to write an [enhancement request](https://github.com/kubernetes/enhancements/) to track the improvement here.

Agreed. There's two steps here:

1. Expose fieldSelector / labelSelector info to the authorizer layer. This will involve changing the requestinfo / authorizerattributes / subjectaccessreview APIs. I think @deads2k had interest in exposing selector information to the authorizer layer, so he may be a good resource here. This would need a design, but is a backwards-compatible addition, so could be easier to do.
2. Tighten the node authorizer to enforce use of node-scoped fieldSelectors when listing/watching nodes/pods. This would require verifying that all supported kubelet versions already limit themselves to listing/watching with these selectors, and solving the issue with `get pod $name`, where the request doesn't have enough information to know if the pod belongs to the node. That would also need a design describing how the `get pod` issue is solved and demonstrating the change is safe with supported kubelets.

File: plugin/pkg/auth/authorizer/node/node_authorizer.go [hunk 1]
```diff
@@ -141,6 +183,20 @@ func (r *NodeAuthorizer) Authorize(ctx context.Context, attrs authorizer.Attribu
 	return authorizer.DecisionNoOpinion, "", nil
 }
 
+// checkListRequest authorizes list requests to objects of a specified types if they are related to the specified node
+func (r NodeAuthorizer) checkListRequest(attrs authorizer.Attributes, nodeName string) (authorizer.Decision, string, error) {
```
[the remark this answers is not in the document]
Comment (sunnoy) [reply plugin/pkg/auth/authorizer/node/node_authorizer.go]: modified

**Comment by sunnoy** [comment]
> This PR _does_ introduce a user facing change. We should only make it:
> 
> * behind a feature gate
> * with a changelog entry
> 
> You may need to write an [enhancement request](https://github.com/kubernetes/enhancements/) to track the improvement here.

creates the issue finished

https://github.com/kubernetes/enhancements/issues/4130

**Comment by sunnoy** [comment]
> > This PR _does_ introduce a user facing change. We should only make it:
> > 
> > * behind a feature gate
> > * with a changelog entry
> > 
> > You may need to write an [enhancement request](https://github.com/kubernetes/enhancements/) to track the improvement here.
> 
> Agreed. There's two steps here:
> 
> 1. Expose fieldSelector / labelSelector info to the authorizer layer. This will involve changing the requestinfo / authorizerattributes / subjectaccessreview APIs. I think @deads2k had interest in exposing selector information to the authorizer layer, so he may be a good resource here. This would need a design, but is a backwards-compatible addition, so could be easier to do.
> 2. Tighten the node authorizer to enforce use of node-scoped fieldSelectors when listing/watching nodes/pods. This would require verifying that all supported kubelet versions already limit themselves to listing/watching with these selectors, and solving the issue with `get pod $name`, where the request doesn't have enough information to know if the pod belongs to the node. That would also need a design describing how the `get pod` issue is solved and demonstrating the change is safe with supported kubelets.


For 1: I don't see the involvement of subjectaccessreview APIs in the nodeAuthorize implementation before the modification, do you mean to add the involvement of subjectaccessreview APIs now?
 
For 2: do need to review the behavior of supported kubelet versions, this will be continued later on

File: staging/src/k8s.io/apiserver/pkg/authorization/authorizer/interfaces.go [hunk 2]
```diff
@@ -62,6 +62,9 @@ type Attributes interface {
 
 	// GetPath returns the path of the request
 	GetPath() string
+
+	// GetQuery returns the Query of the request
+	GetQuery() string
```
[the remark this answers is not in the document]
Comment (sunnoy) [reply staging/src/k8s.io/apiserver/pkg/authorization/authorizer/interfaces.go:67]: Can you give a detailed description, now do not quite understand how to node authentication in the AttributesRecord and RequestInfo and SubjectAccessReview API combined use of the

**Comment by bart0sh** [comment]
@liggitt @sunnoy It looks like this PR is stuck somehow. Can we agree what needs to be done to move it forward?

**Comment by sunnoy** [comment]
> 
Understood. I'll take care of it.

**Comment by sunnoy** [comment]
> There's an enhancements issue open for this PR at [kubernetes/enhancements#4130](https://github.com/kubernetes/enhancements/issues/4130) but no design yet. Before this PR moves forward, there needs to be a [design marked as "implementable"](https://github.com/kubernetes/enhancements/tree/master/keps/#kubernetes-enhancement-proposals-keps) working through the aspects mentioned in [#119471 (comment)](https://github.com/kubernetes/kubernetes/pull/119471#issuecomment-1643970509):
> 
> 1. Addition of labelSelector / fieldSelector attributes into the information exposed to the authorization subsystem (including the SubjectAccessReview API). This aspect is also desired for other authz proposals (like the [ReferenceGrant](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/3766-referencegrant/README.md) proposal which is in design).
> 2. Tightening node authorization to make use of those attributes to only allow access to objects the node should have. Behavior of past kubelet versions, use of kubelet credentials by non-kubelet clients, and ability to filter objects to node-scope using label/field selectors are all important aspects to consider in the design.
> 
> That said, for 1.29, review bandwidth on sig-auth designs/implementations is currently consumed by existing features already in the pipeline. For details on what sig-auth is working on in 1.29, see https://docs.google.com/document/d/1sY8fRyRtk4eG9R439z5ao5i9bFuuxilS03XaNlqoni0/edit (join https://groups.google.com/g/kubernetes-sig-auth for read access) and the [discussion in last week's sig-auth meeting](https://www.youtube.com/watch?v=RID_qxdhCgU&list=PL69nYSiGNLP0VMOZ-V7-5AchXTHAQFzJw&index=2).


Code has been submitted and design has been refined

**Comment by dims** [comment]
@sunnoy can you please sign the CLA? see notes from the bot above. We can't really proceed without it.

**Comment by sunnoy** [comment]
> @sunnoy can you please sign the CLA? see notes from the bot above. We can't really proceed without it.

I've fixed this CLA problem.

**Pull request closed without merging**